src/Middleware/RequestQueue/src/SemaphoreWrapper.cs

```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.RequestQueue
+{
+    public class SemaphoreWrapper
+    {
+        private static SemaphoreSlim semaphore;
+
+        public SemaphoreWrapper(int qlength)
+        {
+            semaphore = new SemaphoreSlim(qlength);
+        }
+
+        public Task EnterQueue()
+        {
+            return semaphore.WaitAsync();
+        }
+
+        public void LeaveQueue()
+        {
+            semaphore.Release();
+        }
+
+        public int SpotsLeft
```

Comment:
"Spots" is a pretty casual term for a property name IMO. I'd suggest something like this:

```suggestion
        public int Count
```

or

```suggestion
        public int QueueLength
```